src/IntersectionObserver.js

```
@@ -69,112 +70,103 @@ export default class IntersectionObserver extends React.Component {
+        this.unobserve(this.targetNode);
+    };
 
-    componentDidUpdate(prevProps) {
-        const propsChanged = observerProps.some(prop => shallowCompare(this.props[prop], prevProps[prop]));
+    getSnapshotBeforeUpdate(prevProps) {
+        this.prevTargetNode = this.targetNode;
 
-        if (propsChanged) {
-            this.unobserve();
+        const relatedPropsChanged = observableProps.some(prop => shallowCompare(this.props[prop], prevProps[prop]));
+        if (relatedPropsChanged) {
+            if (this.prevTargetNode) {
+                if (!prevProps.disabled) {
+                    this.unobserve(this.prevTargetNode);
+                }
+            }
         }
 
-        if (this.targetChanged || propsChanged) {
-            if (!this.props.disabled) {
-                this.observe();
+        return relatedPropsChanged;
+    }
+
+    componentDidUpdate(_, __, relatedPropsChanged) {
+        let targetNodeChanged = false;
+        // check if we didn't unobserve previously due to a prop change
+        if (!relatedPropsChanged) {
+            targetNodeChanged = this.prevTargetNode !== this.targetNode;
+            // check we have a previous node we want to unobserve
+            if (targetNodeChanged && this.prevTargetNode != null) {
+                this.unobserve(this.prevTargetNode);
             }
         }
+
+        if (relatedPropsChanged || targetNodeChanged) {
+            this.observe();
+        }
+    }
+
+    componentDidMount() {
+        this.observe();
     }
 
     componentWillUnmount() {
-        this.unobserve();
+        if (this.targetNode) {
+            this.unobserve(this.targetNode);
+        }
     }
 
     render() {
-        this.renderedTarget = this.target; // this value is null on the first render
+        const { children } = this.props;
 
-        return React.cloneElement(React.Children.only(this.props.children), {
-            ref: this.handleNode,
-        });
+        return children != null
```

@Rendez not sure about side effects, but would `React.isValidElement` work there?

---

It's probably a good idea, specially considering that `isValidElement` unlike the current check, won't skip booleans or numbers, or any other possible primitive, or even non-primitive types. My only concern is that the user won't receive errors, but it will fail silently:
```js
<Observer onChange={...}>{customCheck && anotherCustomCheck}</Observer> // checks evaluate to true (without the user realizing it).
```